### PR TITLE
feat: transformers for windows dotnet 4.x applications

### PR DESCRIPTION
**Pull request by seshapad**
Signed-off-by: seshapad <seshapad@in.ibm.com>

### CODE REVIEW

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+			}
+		}
+	}
+
+	if err := solFileScanner.Err(); err != nil {
+		return nil, fmt.Errorf("Could not parse the solution file: %s", err)
+	}
+
+	for i, c := range projectPaths {
+		c = strings.Replace(c, "\"", "", -1)
```
Comment (ashokponkumar) [inline internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go:106]: In the regex above itself we can match without the "?

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		c = strings.Replace(c, "\\", "/", -1)
```
Comment (ashokponkumar) [inline internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go:107]: Won't this break when running in a windows machine?

Also consider using https://pkg.go.dev/path/filepath#FromSlash and https://pkg.go.dev/path/filepath#ToSlash

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/webappdockerfilegenerator.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		appName: []transformertypes.TransformerPlan{{
+			Mode:              t.Config.Spec.Mode,
+			ArtifactTypes:     []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			BaseArtifactTypes: []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			Paths: map[string][]string{
+				artifacts.ProjectPathPathType: {dir},
+			},
+			Configs: map[string]interface{}{
+				artifacts.DockerfileTemplateConfigConfigType: map[string]interface{}{
+					"Ports":            ports,
```
Comment (ashokponkumar) [inline internal/transformer/classes/generators/dockerfilegenerators/dotnet/webappdockerfilegenerator.go:141]: Might be better to move the port extraction to transform step

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Few more suggestions..

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	return t.Config, t.Env
+}
+
+// BaseDirectoryDetect runs detect in base directory
+func (t *WinConsoleAppDockerfileGenerator) BaseDirectoryDetect(dir string) (namedServices map[string]transformertypes.ServicePlan, unnamedServices []transformertypes.TransformerPlan, err error) {
+	return nil, nil, nil
+}
+
+// parseAppConfig parses the application config
+func (t *WinConsoleAppDockerfileGenerator) parseAppConfig(baseDir string) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+			Mode:              t.Config.Spec.Mode,
+			ArtifactTypes:     []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			BaseArtifactTypes: []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			Paths: map[string][]string{
+				artifacts.ProjectPathPathType: {dir},
+			},
+			Configs: map[string]interface{}{
+				artifacts.DockerfileTemplateConfigConfigType: map[string]interface{}{
+					"Ports":            ports,
+					"baseImageVersion": dotnet.DefaultBaseImageVersion,
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go:219]: Yes. That is the latest image available for <5 projects.

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+func parseSolutionFile(inputPath string) ([]string, error) {
+	solFile, err := os.Open(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open the solution file: %s", err)
+	}
+	defer solFile.Close()
+
+	solFileScanner := bufio.NewScanner(solFile)
+
+	r := regexp.MustCompile(dotnet.ProjBlock)
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go]: I removed this regex altogether.

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+	csr := regexp.MustCompile(dotnet.CsProj)
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go:81]: Instead I used CsProj as the pattern to look for. This change should be seen in conjunction with the above comment:
https://github.com/konveyor/move2kube/pull/578#discussion_r687516351

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+			if len(tokens[1]) > 0 {
+				values := strings.Split(tokens[1], ",")
+				for _, v := range values {
+					if csr.MatchString(v) {
+						projectPaths = append(projectPaths, v)
+					}
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go (hunk 1 above)
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go:88]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/silverlightwebappdockerfilegenerator.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+		appName: []transformertypes.TransformerPlan{{
+			Mode:              t.Config.Spec.Mode,
+			ArtifactTypes:     []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			BaseArtifactTypes: []transformertypes.ArtifactType{artifacts.ContainerBuildArtifactType},
+			Paths: map[string][]string{
+				artifacts.ProjectPathPathType: {dir},
+			},
+			Configs: map[string]interface{}{
+				artifacts.DockerfileTemplateConfigConfigType: map[string]interface{}{
+					"Ports":   ports,
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/silverlightwebappdockerfilegenerator.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/webappdockerfilegenerator.go (hunk 3 above)
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/webappdockerfilegenerator.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+// DirectoryDetect runs detect in each sub directory
+func (t *WinConsoleAppDockerfileGenerator) DirectoryDetect(dir string) (namedServices map[string]transformertypes.ServicePlan, unnamedServices []transformertypes.TransformerPlan, err error) {
+	dirEntries, err := os.ReadDir(dir)
+	if err != nil {
+		logrus.Errorf("Error while trying to read directory: %s", err)
+		return nil, nil, err
+	}
+	appName := ""
+	ports := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/consoleappdockerfilegenerator.go]: done

File: internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go (hunk 2 above)
Comment (seshapad) [reply internal/transformer/classes/generators/dockerfilegenerators/dotnet/utils.go]: Made it os-independent. But did not ToSlash() as it has some weird separator check that does not work in this context. Please refer to https://golang.org/src/path/filepath/path.go

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**